src/main/java/net/fabricmc/loom/api/mappings/intermediate/IntermediateMappingsProvider.java

```diff
@@ -49,4 +50,21 @@ public abstract class IntermediateMappingsProvider implements Named {
 	 * @throws IOException
 	 */
 	public abstract void provide(Path tinyMappings) throws IOException;
+
+	public final void provideWithContext(Path tinyMappings, Project project) throws IOException {
+		if (this instanceof Contextual ctx) {
+			ctx.provide(tinyMappings, project);
+		} else {
+			this.provide(tinyMappings);
+		}
+	}
+
+	public abstract static class Contextual extends IntermediateMappingsProvider {
+		@Override
+		public final void provide(Path tinyMappings) throws IOException {
+			throw new UnsupportedOperationException();
+		}
+
+		public abstract void provide(Path tinyMappings, Project project) throws IOException;
+	}
```

Comment:
I think these should be internal and not public API so they can be changed at a later date, custom intermediary providers can just use the downloader.

I have a been a little bit cautious about passing `Project`s around freely like this. Its very easy to break the rules around configuration caching. It also makes it a little harder to unit test (in this case thats not an issue though).